docs/src/references/mqtt-topic-guidelines.md

```diff
@@ -0,0 +1,608 @@
+   to the equivalent cloud topic for its twin.
+1. It would be ideal if the context/source of data (tedge device, service or child device)
+   can be understood from the topic directly.
+   For e.g: a topic scheme like `tedge/main/{id}`, `tedge/service/{id}` and `tedge/child/{id}` is more user-friendly
+   than a simpler context agnostic scheme like `tedge/{id}` where `id` can be for any "thing".
+1. Limit the topic levels to 7 as AWS IoT core has a [max limit of 7](https://docs.aws.amazon.com/whitepapers/latest/designing-mqtt-topics-aws-iot-core/mqtt-design-best-practices.html)
+
+## Out of scope
+
+1. Routing different kinds of data to different clouds, e.g: all telemetry to azure and all commands from/to Cumulocity.
+   Even though this requirement is realistic, thin-edge MQTT topics must not be corrupted with cloud specific semantics,
+   and the same requirement will be handled with some external routing mechanism(e.g: routing via bridge topics)
+
+# Proposals
+
+## Proposal 1: Dedicated topics for devices and services
+
+This proposal is built on the following assumptions/constraints:
+* thin-edge can identify the child devices (the nested ones as well), and all the services with unique IDs.
+* All child-devices connected to a parent device will have unique IDs under its parent namespace,
+  but not necessarily across the entire tedge namespace/hierarchy.
+* If all child devices **can not** guarantee that uniqueness across the entire hierarchy,
+  then they are expected to register with thin-edge to get that unique ID generated.
+* Given that unique ID, thin-edge can identify its entire parent lineage and descendent child devices.
+  Thin-edge internally maintains this mapping information as child devices register with it.
+* If all child devices **can** guarantee ID uniqueness across the entire hierarchy,
+  they can skip this registration step.
+* Services need not have unique IDs across the entire hierarchy.
+  They just need to have unique IDs under their parent's namespace.
+
+The proposal is as follows:
+
+Topics to have device id as the top level prefix with distinction on the target: device or service as in:
+
+```
+tedge/<device-id>/<target-type>/...
+```
+
+The following demonstrates the different topic prefix for each of the target types:
+
+|Target|Topic Prefix|
+|------|-------|
+|tedge device|`tedge/main/device/`|
+|service on tedge device|`tedge/main/service/<service-id>/`|
+|child device|`tedge/<child-id>/device/`|
+|service on child device|`tedge/<child-id>/service/<service-id>/`|
+
+Where `main` is used as an alias for the tedge device id.
+
+**Why have the `/device/` subtopic level?**
+
+To make a clear distinction between `device` data and `service` data.
+This distinction makes it easier to subscribe to all device only data excluding the service data
+with simple subscription filters like `tedge/main/device/#`.
+Without that device suffix, something like `tedge/main/#` would include the services' data as well.
```

Comment:
What would be the impacts? I see none.

Reply:
It's purely a matter of convenience/consistency. Having that `device` level makes it easier to filter out all data traffic meant for the device only(excluding services), with just `tedge/main/device/#` as opposed to a multi-topic subscriptions to `tedge/main/telemetry/#` and `tedge/main/commands/#`. If we drop the `telemetry` suffix as well, then the list becomes even bigger. And it is also consistent with the `service` topic level.

That suffix can be dropped if simplicity(smaller topics) trumps the marginal convenience and consistency mentioned above.

Reply:
My first impressions were that the `/device/` could be dropped as the difference in subscription is marginal, however having explicit separation between telemetry and control is usually desired in most cases...

The more I read the topics `tedge/main/device`...and `tedge/child/device`, the more I actually like it because it is very explicit about what it represents, e.g. device data. This makes the user more aware about where the data is attached/assigned to.